mdtoc.go

```diff
@@ -1,7 +1,71 @@
 package mdtoc
 
-import "io"
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+)
 
-func Generate(input io.Reader, output io.Writer) {
-	io.Copy(output, input)
+const headerFormat = "- [%s](#%s)"
+
+func writeHeader(output io.Writer, level int, header string) {
+	// TODO: Handle special characters on header
+	line := fmt.Sprintf(headerFormat, header, strings.ToLower(header))
+	// TODO: Handle output writing errors
+	// TODO: Handle when header has trailing spaces
+	output.Write([]byte(line + "\n"))
+}
+
+func parseHeader(parsed []string) (int, string) {
+	var level int
+	for parsed[level] == "#" {
+		level += 1
+	}
+	return level, strings.Join(parsed[level:], "")
+}
+
+func Generate(input io.Reader, output io.Writer) error {
+	headerStart := []byte("<!-- mdtocstart -->")
+	tocHeader := []byte("# Table of Contents")
+	headerEnd := []byte("<!-- mdtocend -->")
+	scanner := bufio.NewScanner(input)
+	var original bytes.Buffer
+	var wroteHeader bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		_, err := original.Write([]byte(line + "\n"))
+		if err != nil {
+			return err
+		}
+		parsed := strings.Split(line, "#")
```

Comment:
I changed some stuff to add some corner cases from markdown, now I use the splitted string and I split by spaces. It is a little odd but it seems that spaces are required on markdown headers, or they are not headers at all. This way I can easily separate the header marker from the rest of the header. Take a look, it looks like crap. To be honest I'm focusing more on handling a lot of corner cases on the tests and less in the algorithm...it is pretty crappy :-)